GEOT-6755 New factory HTTPClientFactory and library gt-http

roarbra opened the pull request:
The main goal is to create a factory for creating HTTPClient. It will also extract some functionality from gt-main into a new library gt-http. The objective for this is to make it more visible, and include the external library Commons HTTPClient.

See also:

* https://osgeo-org.atlassian.net/browse/GEOT-6755
* https://github.com/geotools/geotools/wiki/HTTPClient-Factory

# Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md)
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [X] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [X] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [X] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [x] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

bradh commented:
```
Error:  Found 1 non-complying files, failing build
Error:  To fix formatting errors, run "mvn fmt:format"
Error:  Non complying file: /home/runner/work/geotools/geotools/modules/library/metadata/src/main/java/org/geotools/util/factory/Hints.java
```

aaime commented:
```
2020-12-09T07:22:52.6386528Z [INFO] --- maven-checkstyle-plugin:3.1.1:check (default) @ gt-main ---
2020-12-09T07:22:56.1058225Z [INFO] There is 1 error reported by Checkstyle 8.29 with /home/runner/work/geotools/geotools/build/qa/checkstyle.xml ruleset.
2020-12-09T07:22:56.1081495Z [ERROR] src/main/java/org/geotools/http/HTTPClientFactory.java:[1] (header) RegexpHeader: Line does not match expected header line of '^/\*\s*$'.
```

Did not look, but guessing this java file is missing the GeoTools copyright header. If so, pick it from another file, and please use 2020 as the year.

aaime commented:
See also discussion on the ticket (linked in PR description) about adding Commons HTTP client dependencies to modules that did not have it yet.

roarbra commented:
I don't know exactly how to accomplish the two remaining checkpoints, but I do consider the code and documentation as finished.

jodygarnett commented on modules/library/http/pom.xml:
```diff
@@ -0,0 +1,63 @@
+        <version>25-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>gt-http</artifactId>
```
Just reviewing now, a shame we did not have this to discuss at our bi-weekly meeting as it is an interesting topic.

I think if we adopt this functionality it would end up going into a core library such as metadata or referencing (rather than introducing new jars / dependencies).

aaime replied:
It was discussed in the open already, bug tracker and list, and we suggested to move it to a separate module, to avoid adding a new dependency (commons http client) to all geotools modules.

roarbra replied:
I feel that this functionality has been somewhat hidden away in Geotools. And as #2804 suggest, there are more usages of http-calls than I've handled.
I think it would've been great if all http traffic is handled likewise throughout Geotools. Then it would be easier to deal with bug's regarding https / proxy, and add functionality like logging and url-controlling (ref. #2804 ).
Whether that would be in a new namespace within gt-main, moved to a different library as gt-referencing or gt-metadata (of which I would prefer gt-metadata), or as a separate module in gt-http isn't up to me to decide.
It's perfectly fine for me to wait for the next bi-weekly meeting.

aaime commented on docs/user/library/index.rst:
```diff
@@ -21,3 +21,4 @@ support for common spatial formats such as shapefiles.
    cql/index
    xml/index
    render/index
+   http/index
```
I'm also confused as to how moving and deprecation have been managed, I'm seeing a mix of classes getting moved, and others getting deprecated, with no other code changes, but having a copy in gt-http (e..g, MockHttpClient). Can you clarify?

roarbra replied:
@aaime Here is a table with the deprecated classes and their replacement. I see from the commit history that it isn't easy to follow along which files went where. Some of the files were also changed before I decided to deprecate them.

Deprecated class | Module (original module) | Replacement (other module)
----------------- | ------------------ | ----------------
org.geotools.data.ows.AbstractHttpClient | gt-http (gt-main) | org.geotools.http.AbstractHttpClient
org.geotools.data.ows.MockHttpClient | gt-http (gt-main) | org.geotools.http.MockHttpClient
org.geotools.data.ows.MockHttpResponse | gt-http (gt-main) | org.geotools.http.MockHttpResponse
org.geotools.data.ows.DelegateHTTPClient | gt-http (gt-main) | org.geotools.http.DelegateHTTPClient
org.geotools.data.ows.DelegateHTTPResponse | gt-http (gt-main) | org.geotools.http.DelegateHTTPResponse
org.geotools.data.ows.HTTPClient | gt-http (gt-main) | org.geotools.http.HTTPClient
org.geotools.data.ows.HTTPResponse | gt-http (gt-main) | org.geotools.http.HTTPResponse
org.geotools.data.ows.LoggingHTTPClient | gt-http (gt-main) | org.geotools.http.LoggingHTTPClient
org.geotools.data.ows.SimpleHttpClient | gt-http (gt-main) | org.geotools.http.LoggingHTTPClient
org.geotools.ows.wms.MultithreadedHttpClient | gt-wms | org.geotools.http.MultithreadedHttpClient (gt-http-commons)
org.geotools.ows.MockHttpClient | gt-wms | org.geotools.http.MockHttpClient
org.geotools.ows.MockHttpResponse | gt-wms | org.geotools.http.MockHttpResponse
org.geotools.ows.wmts.MockHttpClient | gt-wmts | org.geotools.http.AbstractHttpClient
org.geotools.data.mongodb.MockHTTPClient | gt-mongodb | org.geotools.http.MockHttpClient
org.geotools.data.mongodb.MockHttpResponse | gt-mongodb | org.geotools.http.MockHttpResponse
org.geotools.ows.wfs.MultithreadedHttpClient | gt-wfs-ng | org.geotools.http.MultithreadedHttpClient (gt-http-commons)
org.geotools.ows.wfs.AbstractTestHTTPClient | gt-wfs-ng | org.geotools.http.AbstractHttpClient
org.geotools.data.Base64 | gt-main | org.geotools.util.Base64 (gt-metadata)

jodygarnett replied:
Adding this table to the documentation.

aaime reviewed (commented):
Wow lots of files changed. Though most are migrations from old to new classes, and are generally pretty obvious.

The one thing standing out at the core level is the relationship between gt-main and gt-http, and the move of the ows classes, see the comments.

The other bit that worries me a bit are the gt-wfs-ng changes, the refactor of test classes there seems deeper, but there is no maintainer to review the changes. I have very little familiarity with the module (despite what the git history might suggest, just did lots of mechanical cleanups while doing QA rounds in the codebase...)

roarbra commented on modules/library/http/pom.xml:
```diff
@@ -0,0 +1,88 @@
+    <dependencies>
+        <dependency>
+            <groupId>org.geotools</groupId>
+            <artifactId>gt-main</artifactId>
```
@aaime Good point! I didn't see that gt-main could depend on gt-http. That way I wouldn't have to move all those classes in "ows".
There are only one problem SimpleHttpClient is using the class org.geotools.data.Base64 which lives in gt-main. How could I solve that problem? I think SimpleHttpClient should live in gt-http.

aaime replied:
I guess Base64 could be moved in gt-metadata, "org.geotools.util" package, and the existing class can become a deprecated subclass of the one in gt-metadata. gt-http will have to depend on gt-metadata anyways due to logging dependencies.

roarbra replied:
I see that solution. Base64 seems unrelated to the other classes of org.geotools.data, but should that be done in a separate PR?

aaime replied:
Hmmm... the move makes sense in the context of this PR though. I'd make it here and be done with it.

jodygarnett commented:
Trying to write up a code example using the factory, and noticing the multithreaded ``http-commons`` does not provide a factory to create instances.

jodygarnett commented:
This may not work on Java 11 due to split-package:

* gt-http: ``org.geotools.data.ows`` package
* gt-main: ``org.geotools.data.ows`` package

Going to see if I can untangle.

roarbra commented:
Hi,

There are some code examples in the documentation.
The thing is that there are only one factory, but it will give different
clients by the hint HTTP_CLIENT in the call for getClient.

For me it seems easier to specify which client to use rather than heving a
factory for all clients and specifying a factory.

The factory have other functionalities like wrapping the client with
another client which will log all http calls. In that respect there might
be a need for a different factory.

Regards,
Roar Brænden


lør. 16. jan. 2021 kl. 00:23 skrev Jody Garnett <notifications@github.com>:

> Trying to write up a code example using the factory, and noticing the
> multithreaded http-commons does not provide a factory to create instances.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotools/geotools/pull/3256#issuecomment-761249350>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN533QHDRY5WOE3QJWM2ADS2DE7BANCNFSM4USARXDQ>
> .
>

aaime commented:
I'm wainting on this one before finishing and merging https://github.com/geotools/geotools/pull/3311, as the test changes would cause a ton of overlap and thus conflicts. So what are the blockers here:

* Java 11 cannot have the same package in two jars (org.geotools.ows)
* See if the wfs-ng tests still pass, with "-Ponline" provided in the maven call

Anything else?

jodygarnett commented:
> I'm wainting on this one before finishing and merging #3311, as the test changes would cause a ton of overlap and thus conflicts. So what are the blockers here:
> 
> * Java 11 cannot have the same package in two jars (org.geotools.ows)
> * See if the wfs-ng tests still pass, with "-Ponline" provided in the maven call
> 
> Anything else?

I started to untangle the split packages here, will also need a factory for gt-http-commons (factory finder cannot access the constructors directly using reflection need to go through service locator). I keep getting distracted by hints and the logic in the factory to supply wrapping and so forth ... 

https://github.com/roarbra/geotools/pull/1

aaime commented:
@jodygarnett I guess what you mean is, under Java 11 access to the constructor by reflection would not be allowed... unless the jar contents are fully exported, no? Isn't what we are doing here?
But yes, if someone wanted to contribute another http module implementation, and they would not expose the classes, then lookup by factory would be the only way. So I guess we have to go back to the current GeoTools factory based design, with a parameter or a hint that would help the factory recognize it's the one that we'd like to use?

roarbra commented:
@jodygarnett So your suggestion would be to have:
For HTTPFactoryFinder
```code
HTTPFactoryFinder.getClient(); // returning SimpleHttpClient

HTTPFactoryFinder.getClient(params) // returning some other client based on params.get("TYPE")
```

In the HTTPClientFactory we would then have two functions:
```code
HTTPClientFactory.canProcess(params) // to select right factory

HTTPClientFactory.createClient(params) // to create the client with the given params
```

Params could then contain all the keys of HTTPClient getter's, as well as choosing logging.

Does this make sense?

aaime commented:
I was trying to figure out where the split package is occurring, checked out the current status of the branch and had a look:

```
aaime@colossus ~/devel/git-gt/modules ((c070d8830a...)) $ git grep "package org.geotools.data.ows"
library/http/src/main/java/org/geotools/data/ows/DelegateHTTPClient.java:package org.geotools.data.ows;
library/http/src/main/java/org/geotools/data/ows/DelegateHTTPResponse.java:package org.geotools.data.ows;
library/http/src/main/java/org/geotools/data/ows/HTTPClient.java:package org.geotools.data.ows;
library/http/src/main/java/org/geotools/data/ows/HTTPResponse.java:package org.geotools.data.ows;
library/http/src/main/java/org/geotools/data/ows/LoggingHTTPClient.java:package org.geotools.data.ows;
library/http/src/main/java/org/geotools/data/ows/SimpleHttpClient.java:package org.geotools.data.ows;
library/http/src/test/java/org/geotools/data/ows/AbstractHttpClient.java:package org.geotools.data.ows;
library/http/src/test/java/org/geotools/data/ows/MockHttpClient.java:package org.geotools.data.ows;
library/http/src/test/java/org/geotools/data/ows/MockHttpResponse.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/AbstractGetCapabilitiesRequest.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/AbstractOpenWebService.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/AbstractRequest.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/Capabilities.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/GetCapabilitiesRequest.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/GetCapabilitiesResponse.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/OperationType.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/Request.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/Response.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/Service.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/ServiceExceptionParser.java:package org.geotools.data.ows;
library/main/src/main/java/org/geotools/data/ows/Specification.java:package org.geotools.data.ows;
library/main/src/test/java/org/geotools/data/ows/ServiceExceptionParserTest.java:package org.geotools.data.ows;
```

As far as I can see, everything in ``library/http/src/main/java/org/geotools/data/ows`` is deprecated classes that are left there for backwards compatibility, to allow a round of deprecation in releases before removing the classes?
However Java 11 split package prevents us from doing that. If I understood correctly and that's the case, I see two options:
- Take those deprecated classes back to gt-main, which is now depending on gt-http
- Remove the classes altoghether (yes, sometimes we do breaking changes, if there are no other reasonable options).

I'd go for the first one, but as said, I don't have a complete understanding of what's going on.

roarbra commented:
> As far as I can see, everything in `library/http/src/main/java/org/geotools/data/ows` is deprecated classes that are left there for backwards compatibility, to allow a round of deprecation in releases before removing the classes?

@aaime They were left there because the new HTTPClient and HTTPResponse extends the old one. I will se what happens if I remove that dependency. Suppose I will have to add some methods to avoid breaking changes.

aaime commented:
> @aaime They were left there because the new HTTPClient and HTTPResponse extends the old one. 

Ouch that is bad, a non deprecated class should not depend on a deprecated one. When we move classes around between packages, we often do the opposite: move the class to a new package, then re-create a empty shell of the old one, in its original place, make it extend the class in the new position, and deprecated it. This avoid code duplication, allows users to still benefit from the same functionality, but get a warning at the same time.

roarbra commented:
> 
> Ouch that is bad, a non deprecated class should not depend on a deprecated one. When we move classes around between packages, we often do the opposite: move the class to a new package, then re-create a empty shell of the old one, in its original place, make it extend the class in the new position, and deprecated it. This avoid code duplication, allows users to still benefit from the same functionality, but get a warning at the same time.

@aaime I've turned it around. And it seems like I'm left with one breaking change. Causing trouble in Geoserver.

For instance in org.geoserver.catalog.ResourcePool:
```code
    class WMSCache extends CatalogResourceCache<String, WebMapServer> {

        @Override
        protected void dispose(String key, WebMapServer server) {
            HTTPClient client = server.getHTTPClient();
            if (client instanceof Closeable) {
                // dispose the client, and the connection pool hosted into it as a consequence
                // the connection pool additionally holds a few threads that are also getting
                // disposed with this call
                @SuppressWarnings("PMD.CloseResource") // actually closing here
                Closeable closeable = (Closeable) client;
                try {
                    closeable.close();
                } catch (IOException e) {
                    LOGGER.log(
                            Level.FINE,
                            "Failure while disposing the http client for a WMS store",
                            e);
                }
            }
        }
    }
```
The trouble line:
```code
            HTTPClient client = server.getHTTPClient();
```
This is returning the new HTTPClient, while Geoserver ResourcePool are expecting the old one.
How do we work around that one?

aaime commented:
I guess the import in GeoServer needs to be modified accordingly.
Also, this should be considered a breaking change and described (along with similar changes) in the upgrade guide.

jodygarnett commented:
> @jodygarnett So your suggestion would be to have:
> For HTTPFactoryFinder
> 
> ```
> HTTPFactoryFinder.getClient(); // returning SimpleHttpClient
> HTTPFactoryFinder.getClient(params) // returning some other client based on params.get("TYPE")
> ```

If you have need to configure the http client during initialization time that is fine, just don't use reflection to obtain the instance - use the FactoryRegistry. The FactoryRegistry makes use of java ServiceLocator which is checks `META-INF/services/<interface name>` file and allowed to create new instances in cases like this where reflection is not allowed.

It is okay to use hints, FactoryFinder method accept hints and will return a factory configured with those hints if one is available.  The specific factor returned is up to the user configuring GeoTools, I assume we should give the apache commons implementation a higher priority that the java ULR based one (so it is made use of if added to the module path).

aside: I would call the methods `create()` and `create(hints)` since this is a factory method responsible for creating a new object. `getClient` implies there may be a single instance which is shared between threads - which is not what you want.
 
> Params could then contain all the keys of HTTPClient getter's, as well as choosing logging.

If you have a need to preconfigure http client instances during creation then uses the params approach is appropriate.

roarbra commented:
@jodygarnett @aaime 

I think we're close to something here.
- Split package should be fixed.
- wfs-ng tests runs as they used to. All I've touch are running with success.
- There's one HTTPClientFactory for each jar. That creates the HTTPClient.
- To get a client you call HTTPFactoryFinder.getClient

Some open questions:
create() vs getClient() - I see your reasoning, but could HTTPFactoryFinder have other factories than HTTPClientFactory, so should we call it createClient()?
Downstream integration fails - Could this be fixed without Geoserver code change?

I was unfortunate to try untangle the code in FactoryRegistry. That caused a lot of trouble, so I reverted to the file from master.

jodygarnett commented:
> Some open questions:
> create() vs getClient() - I see your reasoning, but could HTTPFactoryFinder have other factories than HTTPClientFactory, so should we call it createClient()?

`createClient()` is fine

> Downstream integration fails - Could this be fixed without Geoserver code change?

It looks like this is a real api change, and something that downstream code such as GeoServer will have to address by changing an import, added this example to the documentation.

> I was unfortunate to try untangle the code in FactoryRegistry. That caused a lot of trouble, so I reverted to the file from master.

Understood, there is one thing that is bugging me ... the point of the factory system is to avoid directly importing an implementation such as MultithreadHTTPClient, but ended up with the following code example:

```java
Hints hints = new Hints(Hints.HTTP_CLIENT, org.geotools.http.commons.MultithreadedHttpClient.class);
HTTPClient http = HTTPFactoryFinder.createClient(hints);
HTTPResponse response = http.get();
```

We should be able to set the `MultithreadedHttpClientFactory` with a higher priority, so it is used if it is available on the class path with no further changes to the code.

roarbra commented:
> We should be able to set the `MultithreadedHttpClientFactory` with a higher priority, so it is used if it is available on the class path with no further changes to the code.

Could we please wait with that one? There are code / config within Geoserver WFS that chooses between Simple / Multithreaded. I think we for now should have Simple as default.

aaime commented:
Indeed we have an explicit choice between the two in GeoServer, and a list of dedicated parameters that only work with the multithreaded client. Generalizing it would require a "DataStoreFactory"-like approach, seems like a significant undertaking, both in GeoTools and in GeoServer.

roarbra commented:
> Indeed we have an explicit choice between the two in GeoServer, and a list of dedicated parameters that only work with the multithreaded client. Generalizing it would require a "DataStoreFactory"-like approach, seems like a significant undertaking, both in GeoTools and in GeoServer.

What I mean is that I will look into it in a later stage. I was wrong when I said it was in Geoserver WFS. It's actually in gt-wfs-ng:

```code
    /**
     * Creates the HttpClient instance used to connect to the WFS service, compatible with the given
     * parameters.
     *
     * @param params wfs service connection parameters
     * @return the HttpClient instance
     */
    public HTTPClient getHttpClient(final Map<String, ?> params) throws IOException {
        final URL capabilitiesURL = URL.lookUp(params);
        final WFSConfig config = WFSConfig.fromParams(params);
        if (config.isUseHttpConnectionPooling() && isHttp(capabilitiesURL)) {
            MultithreadedHttpClient client =
                    (MultithreadedHttpClient)
                            HTTPFactoryFinder.createClient(
                                    new Hints(Hints.HTTP_CLIENT, MultithreadedHttpClient.class));

            client.setReadTimeout(config.getTimeoutMillis() / 1000);
            client.setConnectTimeout(config.getTimeoutMillis() / 1000);
            client.setMaxConnections(config.getMaxConnectionPoolSize());

            return client;
        } else {
            return HTTPFactoryFinder.createClient();
        }
    }
```

The suggestion to make MultithreadedHttpClient default when it's on the class path, would make it the result of the else-part above.

aaime commented:
@roarbra the WMS cascading in GeoServer also uses explicit casts to configure the http client, but it's using an explicit creation:
https://github.com/geoserver/geoserver/blob/fec987ccf67d7f511a733a3a72c2ba909dd8a23b/src/main/src/main/java/org/geoserver/catalog/ResourcePool.java#L1942

That said, if the class GeoServer is using is deprecated, it will break the QA build: the settings are the same as GeoTools, deprecated class usage is forbidden.

aaime commented:
By the way, I'm happy to prepare the corresponding GeoWebCache and GeoServer PRs, when @jodygarnett is satisfied with the changes in here (as hinted before, I don't think scope expansion is appropriate in this PR).

jodygarnett commented:
> I was unfortunate to try untangle the code in FactoryRegistry. That caused a lot of trouble, so I reverted to the file from master.

I have also now spent some time trying to untangle the code in FactoryRegistry, aiming for:
```
    public static synchronized HTTPClientFactory getFactory(Hints hints) {
        final Hints merged = mergeSystemHints(hints);
        return getServiceRegistry().getFactory( HTTPClientFactory.class, null, merged, Hints.HTTP_CLIENT_FACTORY );
    }
```
```
    public void findDefaultFactoryLogging() throws Exception {
        Hints hints = new Hints(Hints.HTTP_LOGGING, "TRUE");
        HTTPClientFactory factory = HTTPFactoryFinder.getFactory(hints);
        assertTrue("factory logging found", factory != null);
        assertTrue(
                "logging implementation hint",
                factory.getImplementationHints().containsKey(Hints.HTTP_LOGGING));

        HTTPClient client = factory.createClient();
        assertTrue(client != null);
        assertTrue(client instanceof LoggingHTTPClient);
    }
```

I have gone through and extended GeoTools Factory, implementing the no argument constructuor, constructor that takes hints and reports back implementation hints used during configuration... 

And it is just not working, the FactoryRegisgtry code is a mess of comments and hard to trace through and debug.

jodygarnett commented:
> By the way, I'm happy to prepare the corresponding GeoWebCache and GeoServer PRs, when @jodygarnett is satisfied with the changes in here (as hinted before, I don't think scope expansion is appropriate in this PR).

I am happy with the work thus far, was trying to be helpful but for some reason I cannot get the factory lookup by hints to work.

aaime commented:
@jodygarnett so do you consider the work good to go or is there anything else mandatory here? Don't leave us hanging :-D

jodygarnett commented:
Okay taking step back, echoing discussion from gitter.

Having to have a direct import of a MultiThreadedHttpClient.class or MultiThreadedHttpClientFactory.class negates the reason to introduce a Factory. If you have a direct import anyways you may as well just call the constructor...

1. If creating a MultiThreadedHttpClient with logging, there is no way to configure the connection pooling. At a minium logger could implement the Wrapper interface.

2. Use `getFactory(Hints)` to access the default HTTPClientFactory, or a specific shared instance if HTTP_CLIENT_FACTORY override used.

3. Make a hint for multithreaded, perhaps the connection pool size? Only the MultiThreadedHttpClientFactory will respond to the Hint?

# minor feedback

The HTTP_LOGGING hint can check its input, a bit odd having both true / false or a charset in the same parameter, may of been more direct to have two parameters. Still not sure what the charset parameter would do.

```java
    public static final Key HTTP_LOGGING =
            new Key(String.class) {
                public boolean isCompatibleValue(final Object value) {
                    if (value != null && value instanceof String) {
                        String string = (String) value;
                        if ("TRUE".equals(string) || "FALSE".equals(string)) {
                            return true;
                        }
                        return Charset.isSupported(string);
                    }
                    return false;
                }
            };
```

roarbra commented:
> Okay taking step back, echoing discussion from gitter.
> 

Gitter??? I've followed up on the e-mail list, JIRA and here at Github. I think we've had a fruitful discussion here at GitHub for a month. That certainly could've involved more people. All of a sudden you're referring to some discussion on gitter. That was odd.


> Having to have a direct import of a MultiThreadedHttpClient.class or MultiThreadedHttpClientFactory.class negates the reason to introduce a Factory. If you have a direct import anyways you may as well just call the constructor...
> 

I totally agree that the code for initializing MultithreadedHttpClient seems awkward, but I don't think you will see that code for so very long. It's just the easy way to switch away from:
```code
new SimpleHttpClient()
new MultithreadedHttpClient()
```

The usual way to get a HTTPClient will be:
```code
HTTPFactoryFinder.createClient();
```

Which client this call will return will be given by the system default hints. For the moment the hints to get a specific client is HTTP_CLIENT or HTTP_CLIENT_FACTORY. If the factory only supports one client, and all known factories do that.
These hints is especially useful for injecting mock client's in tests. Without introducing subclasses and methods just for the sake of the tests. Here is an example from gt-mbstyle:
```code

    @Before
    public void setup() throws Exception {
        Hints.putSystemDefault(Hints.HTTP_CLIENT_FACTORY, SpriteMockHttpClientFactory.class);
    }

    @After
    public void tearDown() throws Exception {
        Hints.removeSystemDefault(Hints.HTTP_CLIENT_FACTORY);
    }
```

It could also be useful if someone want's to use the same HTTP client for all their requests in a java program with GeoTools components, or Geoserver for that case. Then they can start with the argument:
```code
-Dorg.geotools.http.client=org.geotools.http.commons.MultithreadedHttpClient
```

At the other hand if your code have specific needs for a HTTP client, we can introduce hints for that as well. In such situations we will use the call:
```code
HTTPFactoryFinder.createClient(hints)
```
And certainly such a hint could be connection_pool_size, which could then end up with a MultithreadedHttpClient.


> 1. If creating a MultiThreadedHttpClient with logging, there is no way to configure the connection pooling. At a minium logger could implement the Wrapper interface.

Good catch. I will look at how to handle it with the current implementation.

> 2. Use `getFactory(Hints)` to access the default HTTPClientFactory, or a specific shared instance if HTTP_CLIENT_FACTORY override used.

I thought your suggestion to drop the getFactory was a good one, so I don't see the reason to go back at that one. You must bring some examples of what you mean here.

> 3. Make a hint for multithreaded, perhaps the connection pool size? Only the MultiThreadedHttpClientFactory will respond to the Hint?
> 

This is a good suggestion, and @aaime said something similar, but should we do it now. Can't we have it as an issue in JIRA and be done with this PR and move on. It will not be a breaking change after we have committed this PR.


> The HTTP_LOGGING hint can check its input, a bit odd having both true / false or a charset in the same parameter, may of been more direct to have two parameters. Still not sure what the charset parameter would do.
> 
> ```java
>     public static final Key HTTP_LOGGING =
>             new Key(String.class) {
>                 public boolean isCompatibleValue(final Object value) {
>                     if (value != null && value instanceof String) {
>                         String string = (String) value;
>                         if ("TRUE".equals(string) || "FALSE".equals(string)) {
>                             return true;
>                         }
>                         return Charset.isSupported(string);
>                     }
>                     return false;
>                 }
>             };
> ```

Just because there was a constructor in LoggingHttpClient with charset, it seemed reasonable to have it as a hint. It's totally fine for me to skip the charset thing.

roarbra commented:
Thinking about MultithreadedHttpClient and what this really is, I think this should be treated as a new interface HttpClientPool. Since it holds on to a ConnectionPool there are some additional concepts other than getting/setting the pool size.

- **Aquiring / releasing connections.** In MultithreadedHttpClient this is done implicit in the get, post method calls, and then releasing when the response is disposed.
- **Closing the pool when we're done with it.** In the case of MultithreadedHttpClient this isn't done. There is a function shutdown on the connectionManager, but I can't see that being called.

To aquire such a client pool, I think the code should be aware of it and calling explicit:
```code
HTTPFactoryFinder.createClientPool(size)
```

What do you think?

roarbra commented:
@aaime
I've introduced the interface HTTPConnectionPooling. The name matches what is used in org.geoserver.catalog.HTTPStoreInfo, and the introduction is to target 1) in @jodygarnett last comment.

When setting maxPoolSize we shouldn't cast to MultithreadedHttpClient, but HTTPConnectionPooling. Then even if we've set logging to on, it will work.

aaime commented:
While working on GWC/GS refactors to support this PR (incoming) I've noticed that the HTTPFactoryFinder was actually:
- Not a FactoryFinder, as in, not something used to locate factories at all (hence, the is-a relationship with the base class was not holding)
- But a client finder instead

Hence, I'm proposing to rename it, PR here https://github.com/roarbra/geotools/pull/2

aaime commented:
GeoWebCache and GeoServer changes here. There is a separate commit for the finder rename, so it's easy to back it out if we don't find an agreement on that change:
* https://github.com/GeoWebCache/geowebcache/pull/918
* https://github.com/geoserver/geoserver/pull/4669

aaime commented:
Now, let's go to the bit that @jodygarnett is not happy about, providing the specific desired client class as a hint, here is an example from GeoServer:
https://github.com/geoserver/geoserver/pull/4669/files#diff-c3dfe96a2f25c04c92b32d554682e4b5f713d7ee5bc0072dc1b8dd3d18a07badR1932

First, it's not a regression, the old code was directly instantiating the clients anyways. In addition to that the old code was doing, this one allows to add logging by setting a system hint, which the old code was not dealing with AFAIK.

Regardless, I believe what @jodygarnett would like, is an ability to migrate clients without having to change user code, like we did for gt-wfs when gt-wgs-ng showed up, or in a similar way, on the switch between gt-postgis and gt-jdbc-postgis.
The case will eventually show up, the day a multithreaded http client is either based on JDK 11 internals, or we switch to HTTP components, so, it's more than just an hyphotesis too.

In datastores, we have a "type" parameter that allows the clients to specify which type of database they want to connect to (because most of the other connection params are shared between databases). So when we did the above switches, we just removed the old module from the classpath, and made the new one react to the same type parameter value. I guess we want to have the same here.

Can we draw a direct parallel here? What about having Hints.HTTP_CLIENT be an Object, accepting both String and Class which could be:
* "POOLING" for any pooling client, the expected result is going to be a HTTPConnectionPooling instance
* "DEFAULT" (or un-specified) if we don't care
* An actual class object, if we want an exact client (retaining the ability for dynamic log wrapping)?

If the client says "POOLING", it's either going to get one, based on what's available in the classpath, or an exception, if no implementation is found. This should allow to switch the implementation of the pooling client from one library to another, without disrupting user code.

Ah, one more thing... "HTTPConnectionPooling" means nothing to me, could we rename it to "HTTPPoolingClient", and also make it a Closeable? If you agree, I'm happy to work on that change too.

aaime commented:
Actually scratch the string values, it can be done with just class values:

* ``HTTPConnectionPoolingc.class`` for any pooling client
* An actual class object, if we want an exact client (retaining the ability for dynamic log wrapping)?
* No value for whatever client (allowing defaults set in the hints)

roarbra commented:
I took another look at org.geotools.data and realized that each factory should have it's own Finder, so I've change HTTPFactoryFinder to HTTPClientFinder.

> Ah, one more thing... "HTTPConnectionPooling" means nothing to me, could we rename it to "HTTPPoolingClient", and also make it a Closeable? If you agree, I'm happy to work on that change too.

That name was taken from Geoserver isConnectionPooling, and MultihtreadedHttpClient does actually have a connection pool rather than a client pool.

I've made it Closeable.


The last point is this initialization of different client's. I commented earlier that I would rather have this as a new proposal / PR, and I stand by that, but if you both feel this must be part of this PR then I will look into it.

aaime commented:
@roarbra a PR with negative comments cannot be merged. We don't do "later" on those, as there is no way to ensure it's going to happen (over the 20 years life of this project, it happened too many times that "later" became "never", and we had to use our spare time to amend, or just live with the issues left... thus the hard line now).

Sorry it's taking so long, I understand it's frustrating... personally I have [large PRs](https://github.com/geotools/geotools/pull/3311) that used several days of my Christmas vacations waiting specifically on this one to be completed and merged, so at least you're in good company ;-)

roarbra commented:
Here is my suggestion. What if we look at HTTPConnectionPooling as a form of HTTPBehavior we want of the client. Others might be HTTPProxy and HTTPGzip.

Then we could add a vararg with HTTPBehavior to the createClient:
```code
HTTPClient createClient(Class<? extends HTTPBehavior> ...behaviors) {

}

HTTPClientFinder.createClient(HTTPConnectionPooling.class)
```

Within HTTPFactoryFinder we should have some code that looks up System.getProperty("http.proxyHost") and if that's set it will add a behavior of HTTPProxy.

I'm still struggling whit the functions HTTPClient.setTryGzip and HTTPClient.isTryGzip. This is certainly a behavior, but when would anyone be specific about using or not using it. Isn't it just a nice to have when both client / server supports it?

@jodygarnett Any feedback on the naming HTTPBehavior would be welcomed.

In this way we will not interfere with the MockHttpClient injection.

jodygarnett commented:
I have some time to look at this again tomorrow, where we at?

roarbra commented:
> I have some time to look at this again tomorrow, where we at?

That would be great. The situation now is:

- The finder is called HTTPClientFinder
- The method createClient takes a varargs of HTTPBehavior
- HTTPBehavior is an interface with two known sub interfaces: HTTPConnectionPooling, HTTPProxy
- HTTP_LOGGING takes true or false as value

The documentation and proposal are updated. The code is implemented and tested.

aaime commented:
The GWC and GS PRs are not aligned mind, so please don't merge until they get fixed (the renaming of HTTPConnectioPooling was not accepted, the sites creating the connection pooling client need to be updated).

jodygarnett commented:
> The documentation and proposal are updated. The code is implemented and tested.

Thanks I like it, reviewing locally and then merge :)

aaime commented:
@jodygarnett yes, going to update the GWC/GS follow up PRs, test some wms/wfs cascading, and then squash-merge

aaime commented:
Merged as https://github.com/geotools/geotools/commit/1fed12f88ab0c9d4ef92aa9cf73af8f94d5c1291

roarbra commented:
Thanks!

aaime commented:
Whops, looks like this change introduced a regression: https://osgeo-org.atlassian.net/browse/GEOT-6850
